### ISSUE TITLE
Another 'git lfs pull' sets all lfs files to modified and changes from lfs to git

### ISSUE THREAD

**Issue by dzieciou** [git-lfs/git-lfs#4913]
**Describe the bug**

When I run `git lfs pull `, all LFS-stored files are marked as modified.

**To Reproduce**

This is happening on our GitHub Enterprise instance, so I cannot share whole repo, but here are the steps

1. Clone repo with some LFS-stored files
    ```
   git clone .../project.git
   cd project/
   ```
2. I make sure there's nothing to be commited for LFS
    ```
    $ git lfs status

    On branch develop

    Objects to be committed:

    Objects not staged for commit:
    ```
3. Then I pull LFS files content and check the status again:
    ```
    $ git lfs pull
    $ git lfs status
    On branch develop

    Objects to be committed:

            notebooks/xyz.ipynb (LFS: e9c3c9b -> Git: e9c3c9b)
            notebooks/something else.ipynb (LFS: 3a9458b -> Git: 3a9458b)
            ...
     ```

**Expected behavior**

`git lfs pull` should not modify existing files status.

**System environment**

Ubuntu 20.04.3 LTS

Output of ` git config -l`:
```
filter.lfs.clean=git-lfs clean -- %f
filter.lfs.smudge=git-lfs smudge -- %f
filter.lfs.process=git-lfs filter-process
filter.lfs.required=true
user.name=Dzieciou
user.email=dzieciou@myorg.com
diff.jupyternotebook.command=git-nbdiffdriver diff
merge.jupyternotebook.driver=git-nbmergedriver merge %O %A %B %L %P
merge.jupyternotebook.name=jupyter notebook merge driver
difftool.nbdime.cmd=git-nbdifftool diff "$LOCAL" "$REMOTE" "$BASE"
difftool.prompt=false
mergetool.nbdime.cmd=git-nbmergetool merge "$BASE" "$LOCAL" "$REMOTE" "$MERGED"
mergetool.prompt=false
credential.helper=cache --timeout=300
filter.lfs.clean=git-lfs clean -- %f
filter.lfs.smudge=git-lfs smudge -- %f
filter.lfs.process=git-lfs filter-process
filter.lfs.required=true
core.repositoryformatversion=0
core.filemode=true
core.bare=false
core.logallrefupdates=true
remote.origin.url=git@github.myorg.net:Product-Management/project.git
remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
branch.develop.remote=origin
branch.develop.merge=refs/heads/develop
```

**Output of `git lfs env`**

```
$ git lfs env
git-lfs/2.13.3 (GitHub; linux amd64; go 1.16.2)
git version 2.33.0

Endpoint=https://github.myorg.net/Product-Management/project.git/info/                                                                                                             lfs (auth=none)
  SSH=git@github.myorg.net:Product-Management/project.git
LocalWorkingDir=/home/dzieciou/projects/lfs/project
LocalGitDir=/home/dzieciou/projects/lfs/project/.git
LocalGitStorageDir=/home/dzieciou/projects/lfs/project/.git
LocalMediaDir=/home/dzieciou/projects/lfs/project/.git/lfs/objects
LocalReferenceDirs=
TempDir=/home/dzieciou/projects/lfs/project/.git/lfs/tmp
ConcurrentTransfers=8
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
LfsStorageDir=/home/dzieciou/projects/lfs/project/.git/lfs
AccessDownload=none
AccessUpload=none
DownloadTransfers=basic,lfs-standalone-file
UploadTransfers=basic,lfs-standalone-file
GIT_EXEC_PATH=/usr/lib/git-core
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"
```

**Additional context**

I followed similar issue #2435 `.gitattributes` seems correct:
```
*.ipnyb filter=lfs diff=lfs merge=lfs -text
```
And the configuration as well:
```
$ git config filter.lfs.process
git-lfs filter-process
```

The only thing that I could think of was that at some point in I moved tracked LFS files from one repository to another.  Could that be a reason?

**Comment by bk2204** [comment]
Hey,

There are a couple pieces of information that would help us track down what's going on.

1. Can you provide the output of `git show HEAD:notebooks/xyz.ipynb` after a fresh clone?
2. Can you tell us whether the files are full files or pointer directly after a clone?
3. Can you provide the output of `git diff -- notebooks/xyz.ipynb` when you see the files are modified?

**Comment by bk2204** [comment]
Thanks for that information.  What does `git status` show?  Do you see different output if you run `git status` before `git lfs status`?  I'm trying to determine if we're simply failing to properly update the index here (which shouldn't be the case, but is possible), or if there's something actually different on disk from what we're expecting.

**Comment by Hackmancoltaire** [comment]
I am having very similar behavior under Ubuntu 20.04.4 LTS. I do a `git lfs clone` or a `git clone` and then a `git lfs pull` and all of my tracked files show as modified. I've been scouring all the other issues related to this and can't find a fix. I get the same results as @dzieciou previous comment.

Config:
```
filter.lfs.clean=git-lfs clean -- %f
filter.lfs.smudge=git-lfs smudge -- %f
filter.lfs.process=git-lfs filter-process
filter.lfs.required=true
filter.lfs.clean=git-lfs clean -- %f
filter.lfs.smudge=git-lfs smudge -- %f
filter.lfs.process=git-lfs filter-process
filter.lfs.required=true
core.repositoryformatversion=0
core.filemode=true
core.bare=false
core.logallrefupdates=true
remote.origin.url=https://[REDACTED].git
remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
branch.master.remote=origin
branch.master.merge=refs/heads/master
lfs.repositoryformatversion=0
lfs.https://[REDACTED]/info/lfs.locksverify=true
```

Attributes:
```
*.sldprt filter=lfs diff=lfs merge=lfs -text
*.sldasm filter=lfs diff=lfs merge=lfs -text
*.slddrw filter=lfs diff=lfs merge=lfs -text
*.mov filter=lfs diff=lfs merge=lfs -text
*.wmv filter=lfs diff=lfs merge=lfs -text
*.mp4 filter=lfs diff=lfs merge=lfs -text
*.mpg filter=lfs diff=lfs merge=lfs -text
*.m4v filter=lfs diff=lfs merge=lfs -text
*.avi filter=lfs diff=lfs merge=lfs -text
*.mkv filter=lfs diff=lfs merge=lfs -text
*.psd filter=lfs diff=lfs merge=lfs -text
*.rar filter=lfs diff=lfs merge=lfs -text
*.zip filter=lfs diff=lfs merge=lfs -text
*.stl filter=lfs diff=lfs merge=lfs -text
*.exe filter=lfs diff=lfs merge=lfs -text
*.bin filter=lfs diff=lfs merge=lfs -text
```

Git LFS Env:
```
git-lfs/3.1.2 (GitHub; linux amd64; go 1.17.6)
git version 2.35.1

Endpoint=https://[REDACTED].git/info/lfs (auth=none)
LocalWorkingDir=/mnt/storage/pdmtogit/resources/git
LocalGitDir=/mnt/storage/pdmtogit/resources/git/.git
LocalGitStorageDir=/mnt/storage/pdmtogit/resources/git/.git
LocalMediaDir=/mnt/storage/pdmtogit/resources/git/.git/lfs/objects
LocalReferenceDirs=
TempDir=/mnt/storage/pdmtogit/resources/git/.git/lfs/tmp
ConcurrentTransfers=8
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
LfsStorageDir=/mnt/storage/pdmtogit/resources/git/.git/lfs
AccessDownload=none
AccessUpload=none
DownloadTransfers=basic,lfs-standalone-file,ssh
UploadTransfers=basic,lfs-standalone-file,ssh
GIT_EXEC_PATH=/usr/lib/git-core
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"
```

**Comment by Hackmancoltaire** [comment]
I think I might have found the solution for my problem. Running the following command resolved the issue.

`git config --global core.ignorecase true`

Found the answer here:
https://stackoverflow.com/a/69440512

Looks like LFS attributes are case sensitive?

**Comment by dzieciou** [comment]
> Thanks for that information. What does git status show? Do you see different output if you run git status before git lfs status? I'm trying to determine if we're simply failing to properly update the index here (which shouldn't be the case, but is possible), or if there's something actually different on disk from what we're expecting.

It looks like the output of `git status` is same as output of following `git lfs status`, both for fresh clone of repository and after running `git lfs pull`.

```
$ git status
On branch develop
Your branch is up to date with 'origin/develop'.

nothing to commit, working tree clean

$ git lfs status
On branch develop
Objects to be pushed to origin/develop:


Objects to be committed:


Objects not staged for commit:
```

After `git lfs pull`:
```
$ git lfs pull

$ git status
On branch develop
Your branch is up to date with 'origin/develop'.

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        modified:   notebooks/xyz.ipynb
        modified:   notebooks/something else.ipynb 
		
$ git lfs status
On branch develop
Objects to be pushed to origin/develop:


Objects to be committed:

        notebooks/xyz.ipynb (LFS: e9c3c9b -> Git: e9c3c9b)
        notebooks/something else.ipynb (LFS: 3a9458b -> Git: 3a9458b)
```

**Comment by dzieciou** [comment]
> I think I might have found the solution for my problem. Running the following command resolved the issue.
> 
> `git config --global core.ignorecase true`
> 
> Found the answer here: https://stackoverflow.com/a/69440512
> 
> Looks like LFS attributes are case sensitive?

In my case it didn't help.

**Comment by bk2204** [comment]
> I think I might have found the solution for my problem. Running the following command resolved the issue.
> 
> `git config --global core.ignorecase true`
> 
> Found the answer here: https://stackoverflow.com/a/69440512
> 
> Looks like LFS attributes are case sensitive?

Git is generally case sensitive, and Git LFS has to follow whatever Git does.  `core.ignorecase` is set on a per-repository basis based on whether Git thinks the repo lives on a case-sensitive or case-insensitive file system.  It's not supposed to be set to something different than Git thinks it is, because that will end up breaking things.  You are very definitely not supposed to set it globally.

It's likely your repository was created on Windows or a case-insensitive macOS and wasn't set up properly to work on a Linux system, since you probably have files that differ from the patterns in case.  You'd be better off talking to the author of the repository and fixing the case of the files such that they're all consistent with the patterns in `.gitignore`.

**Comment by dzieciou** [comment]
```
$ cat .gitattributes
*.ipnyb filter=lfs diff=lfs merge=lfs -text
```
That's the only line there.

Here's `git diff` for regular files:
```
$ echo "xyz" > x.txt
$ git add x.txt
$ git diff -- x.txt

diff --git a/x.txt b/x.txt
index e69de29..cd470e6 100644
--- a/x.txt
+++ b/x.txt
@@ -0,0 +1 @@
+xyz
```
Aparently, for LFS-matching files, there's not diff returned:
```
$ echo "xyz" > x.ipynb
$ git add x.ipynb
$ git diff -- x.ipynb | wc
      0       0       0
```
And git status doesn't say that the content of `x.ipynb` has been modified, only that it has been added:
```
$ git status
On branch develop
Your branch is up to date with 'origin/develop'.

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        new file:   x.ipynb
        new file:   x.txt

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
```

**Comment by bk2204** [comment]
Hey,

What I'm really looking for is the diff output after you do a `git lfs pull`; that is, just a generic `git diff`.  Git thinks some of those files are modified, and I want to know what it thinks is modified about them when they're in the working tree.

I expect a lack of diff output when you do `git add x.ipynb`, because added items would need `git diff --cached` to see the difference.

**Comment by bk2204** [comment]
Hey,

This is consistent with what `git lfs status` is telling us: that the file is being converted from a pointer file to a plain Git file, which isn't correct.  That would imply that the clean filter isn't working.  Can you run the `git lfs pull` and `git lfs status` commands with `GIT_TRACE=1`?  Also, where in your path is the `git-lfs` binary located (i.e., can you provide the output of `which -a git-lfs`)?

**Comment by bk2204** [comment]
> > That would imply that the clean filter isn't working.
> 
> I've never heard of that filter. What does it do?

Git LFS operates as a smudge and clean filter.  For files which have `filter=lfs` set in `.gitattributes`, the pointer file is stored in Git and Git invokes Git LFS to smudge the pointer file into a large file.  Similarly, when a file is added to Git or refreshed in the index, Git runs it through Git LFS to clean it into a pointer file.

And I believe I've just found the problem.  Your `.gitattributes` file is misspelled:

```
*.ipnyb filter=lfs diff=lfs merge=lfs -text
```

It should be `*.ipynb`, not `*.ipnyb`.  If you fix that and commit, it should fix the problem.

**Comment by dzieciou** [comment]
That resolves the problem. Thank you for troubleshooting that.

Closing.

**Comment by michelau** [comment]
For those that might find their way here, I had this same problem, and it was also due to an error in `.gitattributes`.  The difference is that the `.gitattributes` was generated by https://github.com/rtyley/bfg-repo-cleaner as part of its Git LFS conversion feature.

I passed the following to BFG:
`--convert-to-git-lfs "{*.jpg}"`

This matched all of the `*.jpg` files as expected, but also resulted in the following invalid `.gitattributes`:
`{*.jpg} filter=lfs diff=lfs merge=lfs -text`

After changing it manually to the following everything works fine:
`*.jpg filter=lfs diff=lfs merge=lfs -text`

**Comment by ashishinhaufe** [comment]
@michelau I have used exactly same tool and facing this problem. For me when i do git lfs pull it shows files with changes. I made changes to .gitattribute file to remove that extra curly braces as suggested in your comment. It still shows binary files changed and staged after running git lfs pull. Did you also faced similar issue ? Do i have to push the .gitattribute file to repo before running git lfs pull ?
